File: distribution/python/dqops/client/models/check_results_overview_data_model.py
import datetime
from typing import Any, Dict, List, Type, TypeVar, Union, cast

from attrs import define as _attrs_define
from attrs import field as _attrs_field
from dateutil.parser import isoparse

from ..models.check_result_status import CheckResultStatus
from ..types import UNSET, Unset

T = TypeVar("T", bound="CheckResultsOverviewDataModel")


@_attrs_define
class CheckResultsOverviewDataModel:
    """
    Attributes:
        check_hash (Union[Unset, int]): Check hash.
        check_category (Union[Unset, str]): Check category name.
        check_name (Union[Unset, str]): Check name.
        comparison_name (Union[Unset, str]): Optional table comparison name for table comparison checks only.
        time_periods (Union[Unset, List[datetime.datetime]]): List of time periods for the results, returned as a local
            time, sorted from the newest to the oldest.
        time_periods_utc (Union[Unset, List[int]]): List of time periods for the results, returned as absolute UTC time.
        executed_at_timestamps (Union[Unset, List[int]]): List of absolute timestamp (UTC) when the check was executed
            or an error was raised.
        time_period_display_texts (Union[Unset, List[str]]): List of time periods, sorted descending, returned as a text
            with a possible time zone.
        statuses (Union[Unset, List[CheckResultStatus]]): List of check severity levels or an error status, indexes with
            the severity levels match the time periods.
        data_groups (Union[Unset, List[str]]): List of data group names. Identifies the data group with the highest
            severity or error result.
        results (Union[Unset, List[float]]): List of sensor results. Returns the data quality result readout for the
            data group with the alert of the highest severity level.
    """

    check_hash: Union[Unset, int] = UNSET
    check_category: Union[Unset, str] = UNSET
    check_name: Union[Unset, str] = UNSET
    comparison_name: Union[Unset, str] = UNSET
    time_periods: Union[Unset, List[datetime.datetime]] = UNSET
    time_periods_utc: Union[Unset, List[int]] = UNSET
    executed_at_timestamps: Union[Unset, List[int]] = UNSET
    time_period_display_texts: Union[Unset, List[str]] = UNSET
    statuses: Union[Unset, List[CheckResultStatus]] = UNSET
    data_groups: Union[Unset, List[str]] = UNSET
    results: Union[Unset, List[float]] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        check_hash = self.check_hash
        check_category = self.check_category
        check_name = self.check_name
        comparison_name = self.comparison_name
        time_periods: Union[Unset, List[str]] = UNSET
        if not isinstance(self.time_periods, Unset):
            time_periods = []
            for time_periods_item_data in self.time_periods:
                time_periods_item = time_periods_item_data.isoformat()

                time_periods.append(time_periods_item)

        time_periods_utc: Union[Unset, List[int]] = UNSET
        if not isinstance(self.time_periods_utc, Unset):
            time_periods_utc = self.time_periods_utc

        executed_at_timestamps: Union[Unset, List[int]] = UNSET
        if not isinstance(self.executed_at_timestamps, Unset):
            executed_at_timestamps = self.executed_at_timestamps

        time_period_display_texts: Union[Unset, List[str]] = UNSET
        if not isinstance(self.time_period_display_texts, Unset):
            time_period_display_texts = self.time_period_display_texts

        statuses: Union[Unset, List[str]] = UNSET
        if not isinstance(self.statuses, Unset):
            statuses = []
            for statuses_item_data in self.statuses:
                statuses_item = statuses_item_data.value

                statuses.append(statuses_item)

        data_groups: Union[Unset, List[str]] = UNSET
        if not isinstance(self.data_groups, Unset):
            data_groups = self.data_groups

        results: Union[Unset, List[float]] = UNSET
        if not isinstance(self.results, Unset):
            results = self.results

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if check_hash is not UNSET:
            field_dict["checkHash"] = check_hash
        if check_category is not UNSET:
            field_dict["checkCategory"] = check_category
        if check_name is not UNSET:
            field_dict["checkName"] = check_name
        if comparison_name is not UNSET:
            field_dict["comparisonName"] = comparison_name
        if time_periods is not UNSET:
            field_dict["timePeriods"] = time_periods
        if time_periods_utc is not UNSET:
            field_dict["timePeriodsUtc"] = time_periods_utc
        if executed_at_timestamps is not UNSET:
            field_dict["executedAtTimestamps"] = executed_at_timestamps
        if time_period_display_texts is not UNSET:
            field_dict["timePeriodDisplayTexts"] = time_period_display_texts
        if statuses is not UNSET:
            field_dict["statuses"] = statuses
        if data_groups is not UNSET:
            field_dict["dataGroups"] = data_groups
        if results is not UNSET:
            field_dict["results"] = results

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        check_hash = d.pop("checkHash", UNSET)

        check_category = d.pop("checkCategory", UNSET)

        check_name = d.pop("checkName", UNSET)

        comparison_name = d.pop("comparisonName", UNSET)

        time_periods = []
        _time_periods = d.pop("timePeriods", UNSET)
        for time_periods_item_data in _time_periods or []:
            time_periods_item = isoparse(time_periods_item_data)

            time_periods.append(time_periods_item)

        time_periods_utc = cast(List[int], d.pop("timePeriodsUtc", UNSET))

        executed_at_timestamps = cast(List[int], d.pop("executedAtTimestamps", UNSET))

        time_period_display_texts = cast(
            List[str], d.pop("timePeriodDisplayTexts", UNSET)
        )

        statuses = []
        _statuses = d.pop("statuses", UNSET)
        for statuses_item_data in _statuses or []:
            statuses_item = CheckResultStatus(statuses_item_data)

            statuses.append(statuses_item)

        data_groups = cast(List[str], d.pop("dataGroups", UNSET))

        results = cast(List[float], d.pop("results", UNSET))

        check_results_overview_data_model = cls(
            check_hash=check_hash,
            check_category=check_category,
            check_name=check_name,
            comparison_name=comparison_name,
            time_periods=time_periods,
            time_periods_utc=time_periods_utc,
            executed_at_timestamps=executed_at_timestamps,
            time_period_display_texts=time_period_display_texts,
            statuses=statuses,
            data_groups=data_groups,
            results=results,
        )

        check_results_overview_data_model.additional_properties = d
        return check_results_overview_data_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/data/checkresults/services/models/CheckResultsOverviewDataModel.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.data.checkresults.services.models;

import com.dqops.checks.CheckTimeScale;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.google.common.collect.Lists;
import lombok.Data;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Check recent results overview. Returns the highest severity for the last several runs.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.LowerCamelCaseStrategy.class)
@Data
public class CheckResultsOverviewDataModel {
    private static final DateTimeFormatter MONTH_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM X", Locale.ROOT);

    @JsonPropertyDescription("Check hash.")
    private long checkHash;

    @JsonPropertyDescription("Check category name.")
    private String checkCategory;

    @JsonPropertyDescription("Check name.")
    private String checkName;

    @JsonPropertyDescription("Optional table comparison name for table comparison checks only.")
    private String comparisonName;

    @JsonPropertyDescription("List of time periods for the results, returned as a local time, sorted from the newest to the oldest.")
    private List<LocalDateTime> timePeriods = new ArrayList<>();

    @JsonPropertyDescription("List of time periods for the results, returned as absolute UTC time.")
    private List<Instant> timePeriodsUtc = new ArrayList<>();

    @JsonPropertyDescription("List of absolute timestamp (UTC) when the check was executed or an error was raised.")
    private List<Instant> executedAtTimestamps = new ArrayList<>();

    @JsonPropertyDescription("List of time periods, sorted descending, returned as a text with a possible time zone.")
    private List<String> timePeriodDisplayTexts = new ArrayList<>();

    @JsonPropertyDescription("List of check severity levels or an error status, indexes with the severity levels match the time periods.")
    private List<CheckResultStatus> statuses = new ArrayList<>();

    @JsonPropertyDescription("List of data group names. Identifies the data group with the highest severity or error result.")
    private List<String> dataGroups = new ArrayList<>();

    @JsonPropertyDescription("List of sensor results. Returns the data quality result readout for the data group with the alert of the highest severity level.")
    private List<Double> results = new ArrayList<>();

    /**
     * Appends a new result. If the time period is not known, a new set of values are added. If the time period matches the most recent
     * time period then the current severity and data stream is replaced if the severity is higher.
     * NOTE: This method must be called in order, adding the most recent time periods first. Following time periods must be descending.
     * @param timePeriod Time period to add.
     * @param timePeriodUtc Time period converted to UTC time (real).
     * @param executedAt Executed at timestamp.
     * @param checkTimeScale Time scale used by the check. For daily checks, the results are truncated.
     * @param severity Check severity.
     * @param actualValue Actual value returned by the sensor.
     * @param dataGroup Data group.
     * @param resultsCount Maximum results count to store. The result is not added if the result count is exceeded.
     */
    public void appendResult(LocalDateTime timePeriod,
                             Instant timePeriodUtc,
                             Instant executedAt,
                             CheckTimeScale checkTimeScale,
                             Integer severity,
                             Double actualValue,
                             String dataGroup,
                             int resultsCount) {
        assert this.timePeriods.size() == 0 || !this.timePeriods.get(this.timePeriods.size() - 1).isAfter(timePeriod);

        if (this.timePeriods.size() == 0) {
            this.timePeriods.add(timePeriod);
            this.timePeriodsUtc.add(timePeriodUtc);
            this.executedAtTimestamps.add(executedAt);
            this.timePeriodDisplayTexts.add(makeTimePeriodDisplayText(timePeriod, timePeriodUtc, checkTimeScale));
            this.statuses.add(CheckResultStatus.fromSeverity(severity));
            this.dataGroups.add(dataGroup);
            this.results.add(actualValue);
            return;
        }

        if (this.timePeriodsUtc.get(this.timePeriodsUtc.size() - 1).equals(timePeriodUtc)) {
            if (severity != 4 && severity > this.statuses.get(this.statuses.size() - 1).getSeverity()) {
                // another result with a higher severity, replacing the current one, we found a bigger issue, but we are ignoring errors
                this.executedAtTimestamps.set(this.statuses.size() - 1, executedAt);
                this.timePeriodDisplayTexts.set(this.statuses.size() - 1, makeTimePeriodDisplayText(timePeriod, timePeriodUtc, checkTimeScale));
                this.statuses.set(this.statuses.size() - 1, CheckResultStatus.fromSeverity(severity));
                this.dataGroups.set(this.dataGroups.size() - 1, dataGroup);
                this.results.set(this.dataGroups.size() - 1, actualValue);
            }
        }
        else {
            if (this.timePeriods.size() >= resultsCount) {
                return;
            }

            this.timePeriods.add(timePeriod);
            this.timePeriodsUtc.add(timePeriodUtc);
            this.executedAtTimestamps.add(executedAt);
            this.timePeriodDisplayTexts.add(makeTimePeriodDisplayText(timePeriod, timePeriodUtc, checkTimeScale));
            this.statuses.add(CheckResultStatus.fromSeverity(severity));
            this.dataGroups.add(dataGroup);
            this.results.add(actualValue);
        }
    }

    /**
     * Creates a time period text that is valid text to be shown.
     * @param timePeriod Time period in local time.
     * @param timePeriodUtc Time period as UTC absolute time.
     * @param checkTimeScale Check time scale (daily, etc) how to format the date or datetime.
     * @return Display text that is usable, optionally has a time zone.
     */
    protected String makeTimePeriodDisplayText(LocalDateTime timePeriod, Instant timePeriodUtc, CheckTimeScale checkTimeScale) {
        Instant timePeriodForcedUtc = timePeriod.toInstant(ZoneOffset.UTC);
        long timeOffsetSeconds = timePeriodForcedUtc.getEpochSecond() - timePeriodUtc.getEpochSecond();
        if (timeOffsetSeconds > 18L * 3600L || timeOffsetSeconds < -18L * 3600L) {
            timeOffsetSeconds = timeOffsetSeconds % (24L * 3600L);
        }

        ZoneOffset zoneOffset = ZoneOffset.ofTotalSeconds((int) timeOffsetSeconds);

        if (checkTimeScale != null) {
            switch (checkTimeScale) {
                case daily:
                    return timePeriod.atOffset(zoneOffset).format(DateTimeFormatter.ISO_OFFSET_DATE)
                            .replace("Z", " UTC");

                case monthly:
                    return timePeriod.atOffset(zoneOffset).format(MONTH_FORMATTER)
                            .replace("Z", "UTC");
            }
        }

        return timePeriod.atOffset(zoneOffset).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
                .replace('T', ' ')
                .replace("Z", " UTC");
    }

    /**
     * Reverses the lists because we are appending the most recent (newest) statuses first, but we want to return the results ordered in chronological order (the most recent is the last in the list).
     */
    public void reverseLists() {
        this.timePeriods = Lists.reverse(this.timePeriods);
        this.timePeriodsUtc = Lists.reverse(this.timePeriodsUtc);
        this.executedAtTimestamps = Lists.reverse(this.executedAtTimestamps);
        this.timePeriodDisplayTexts = Lists.reverse(this.timePeriodDisplayTexts);
        this.statuses = Lists.reverse(this.statuses);
        this.dataGroups = Lists.reverse(this.dataGroups);
        this.results = Lists.reverse(this.results);
    }
}
